Standardize event handler `this` to be the same as JS events, and pre…

cornelius-dol opened the pull request:
Standardize event handler `this` to match JS events; prefer original target nearest node over current target node.

Refer to #240 for rationale.

cornelius-dol committed:
Standardize event handler `this` to match JS events; prefer original target nearest node over current target node.

cornelius-dol committed:
Inline the exec function.

cornelius-dol committed:
Change event handling back to using add/removeEventHandler.

lawrence-dol commented:
@leeoniya : I think the only contentious aspect of this PR is the binding of `this` for parameterized event handlers being `currentTarget` vs. `vm`. I am only too happy to drop that change if it means getting this merged and and build out. 

Currently I have two projects with DOMVM patched manually to fix the add/remove handler issue. Besides that I have this version deployed in a number of quite complex scenarios without any issues -- Normalizing the events to leverage JavaScript's handling has made events hugely easier for me to reason about.

PS: I am sorry if I've created more work for you at this point, not less. Hopefully that will change and future contributions will be simple, surgical improvements. I am looking forward to putting my mind to looking for performance enhancements (by doing less work, not by fiddling with constructs, since the minimizer is going to completely rework the code anyway).

lawrence-dol commented:
@leeoniya Ping!

cornelius-dol committed:
Critical fix for patchEvent bug which occurs when the view data is an array, introduced in `365bafd...`.

Using an array for the DOMVM args in args.concat was important because concat flattens the first level of arrays.
When the view data is an array, the broken code was flattening its elements into additional arguments instead of
passing the data as an array.

lawrence-dol commented:
@le

lawrence-dol commented:
@leeoniya : I am starting to feel very antsy about this change. I have multiple projects all using the test build from this commit. The only problem I've encountered is the mishandling of the view data for which I just pushed a fix.

I confess to feeling troubled that I have production systems dependent on DOMVM and I don't have the ability to get fixes and enhancements released in a reasonable time frame. I am willing to do anything I can to assist, since my only other option is to fork the project and just proceed on my own -- something I am loath to do.

leeoniya commented:
hey @lawrence-dol sorry for the huge delay, i have a ton of work-related stuff on my plate which makes giving thoughtful & thorough reviews rather challenging at the moment. hopefully i will find some time next week.

> since my only other option is to fork the project and just proceed on my own -- something I am loath to do.

i don't see much of a problem with doing this temporarily as long as the PRs are likely to get merged eventually and there is no fundamental disagreement about the changes. i think you've worked enough in this part of the codebase that i won't have major objections so long as you make sure that performance is not affected in a significant way (unless agreed upon) and no memory leaks are introduced.

leeoniya commented on test/src/streams.js:
```diff
@@ -30,7 +30,7 @@ var streamHooks = {
 };
 
-QUnit.module("streams", streamHooks, function() {
+QUnit.skip("streams", streamHooks, function() {
```
this was just failing in your env, right? not sure it should be disabled entirely.

lawrence-dol replied:
Yes, it fails for me when I run the DOMVM tests. I expect all tests to succeed in a project without needing to install anything further than the project and its dependencies. My impression is that they are broken until changes can be made to them.

leeoniya commented on src/view/utils.js:
```diff
@@ -48,4 +48,11 @@ export function getVm(n) {
+
+export function getVnode(el) {
+	el = el || emptyObj;
+	while (el._node == null && el.parent)
```
`el.parent` isn't a thing, did you mean `el.parentNode`?

lawrence-dol replied:
Oh, crap. Good catch! 

That's got me many times because all the other functions for relational nodes in Node do not end with ...Node (firstChild, lastChild, nextSibling, previousSibling). This is why API consistency is so important to me -- especially in a weakly typed or untyped language. It's particularly important in JS where things you reference might be missing and mistyped names aren't caught at the reference point.

lawrence-dol replied:
Resolved by commit 0801af0cd3f9f15bda56deee16a978f033598643.

leeoniya commented on src/view/utils.js:
```diff
@@ -48,4 +48,11 @@ export function getVm(n) {
+}
+
+export function getVnode(el) {
+	el = el || emptyObj;
```
when would `el` be falsey? is there a case where `evt.target` is falsey?

lawrence-dol replied:
Not currently, but I was wanting to continue the discussion about exposing this function as part of the API. We did briefly discuss that as potentially useful at the app level.

Also, I tend to lean a little defensive in the utility function design when the logical expectation of a null input would be a null output, especially if null output is already a possibility.

In this case, I could go either way -- null input is invalid and causes a fast-failure, or it's valid and causes a null output.

lawrence-dol commented:
@leeoniya : Ping!

lawrence-dol commented:
@leeoniya 

Hey I don't want to be a pest, but I have been deploying "3.4.14.alpha" into production since early this year (that being a build with this branch). It's been working perfectly well, and I believe all objections raised in this thread have been resolved. I believe it's past time to create the 3.4.14 build.

FWIW, I have been explaining to my team for a while why it is that I have deployed an "alpha" library to our production software, and it's starting to get a bit awkward. I have championed DOMVM over alternatives like React, in part using the defense that if it's ever abandoned we can simply fork it and take over maintenance... pressure is mounting for me to do exactly that. This would apply to the other pull request I have pending for chaining `config()` and to your fix for capturing render exceptions (branch handle-render-exceptions).

leeoniya commented:
i will look at this tomorrow

leeoniya commented:
@lawrence-dol i pushed some whitespace cleanup (spaces -> tabs) and a fix to get the tests passing (`.concat(args)` works as a multi-push when passed an array, a javascript-ism)

can you verify the latest version in your app?

lawrence-dol commented:
@leeoniya : Seems to be working just fine. I made a minor tweak to avoid cloning the DOMVM args (about which I have mixed feelings, TBH). I'll clean up the history and force push for final review.

lawrence-dol commented:
@leeoniya : We seem to be at odds with leading whitespace being tabs vs. spaces. My editor converts all tabs to spaces on save, since that's what our inhouse conventions require. Do you need me to convert leading tabs to spaces for DOMVM? I don't think Atom can inherently do that on a per-project basis, but I can probably rig something up.

cornelius-dol committed:
spaces -> tabs

lawrence-dol commented:
No issues observed with this build all day, and it's essentially the same code I've been using in several project for several months.

lawrence-dol commented:
(BTW, I switch Atom to use leading tabs if it detects tabs in the existing source file. That should remedy the problem with my edits switch the file/lines to leading spaces.)

lawrence-dol commented:
@leeoniya 

I've cleaned up the history as much as I think is beneficial, and found some places to reduce the code size in the process. But I want to double-check one thing, by bringing it to your attention before we merge this.

The original code for calling the events was:

	if (FEAT_ONEVENT) {
		out2 = vm.onevent(e, node, vm, vm.data, args),
		out3 = onevent.call(null, e, node, vm, vm.data, args);
	}

while the current code is (which I changed):

	if (FEAT_ONEVENT) {
		out2 = vm.onevent(evt, node, vm, vm.data, args),
		out3 =    onevent(evt, node, vm, vm.data, args);
	}

A minor difference to be sure. I am pretty sure I changed it for congruence, but maybe you had a specific reason for explicitly setting the `this` value for the top-level `onevent` handler?

leeoniya commented:
> but maybe you had a specific reason for explicitly setting the this value for the top-level onevent handler?

not that i can recall.

i gotta say i'm not super keen on binding `this` to `e.currentTarget` for congruence, since relying on `this` is generally discouraged and `this` is not available in arrow function handlers. also, any congruence with plain handlers is broken anyways as soon as you have more than 0 parameters, since they get prepended. i would much rather people just explicitly access `e.currentTarget`, e.g.:

https://jsfiddle.net/sfyw15a9/4/

```js
const el = domvm.defineElement;
const cv = domvm.createView;

let handleClick = e => {
  console.log(e.target, e.currentTarget);
};

let handleClickParamd = (...args) => {
  handleClick(args.find(a => a instanceof Event));
};

const View = (vm, data) => {
  return () =>
    el("div", [
      el("div", {onclick: handleClick}, [
        el("input"),
      ]),
      el("div", {onclick: [handleClickParamd, 1, 2]}, [
        el("input"),
      ])
    ]);
};

cv(View).mount(document.body);
```

do you still feel strongly about explicitly binding `this`?

lawrence-dol commented:
For myself _using_ DOMVM I don't care a whit about what `this` is bound to as I never use it. The one time I thought I had to use `this`, which happened to be when I discovered the incongruence, it turned out I was wrong, misunderstanding the salient event property.

But for the sake of the API I just can't find a justification for _not_ binding `this` in the same way that a normal handler would.

lawrence-dol commented:
Thanks!

leeoniya commented:
@lawrence-dol https://www.npmjs.com/package/domvm/v/3.4.14
